File: src/proxy/fake_tls.rs
use hmac::{Hmac, KeyInit, Mac};
use rand::RngExt;
use sha2::Sha256;
use std::collections::VecDeque;
use std::pin::Pin;
use std::task::{Context, Poll};
use std::time::{SystemTime, UNIX_EPOCH};

use tokio::io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt};
use tokio::net::TcpStream;

use crate::proxy::STATS;

type HmacSha256 = Hmac<Sha256>;

pub const TLS_RECORD_HANDSHAKE: u8 = 0x16;
pub const TLS_RECORD_CCS: u8 = 0x14;
pub const TLS_RECORD_APPDATA: u8 = 0x17;

const CLIENT_RANDOM_OFFSET: usize = 11;
const CLIENT_RANDOM_LEN: usize = 32;
const SESSION_ID_OFFSET: usize = 44;
const SESSION_ID_LEN: usize = 32;
const TIMESTAMP_TOLERANCE: i64 = 120;
const TLS_APPDATA_MAX: usize = 16384;

const CCS_FRAME: [u8; 6] = [0x14, 0x03, 0x03, 0x00, 0x01, 0x01];

/// Verify TLS ClientHello with Fake TLS (ee-secret) protocol.
/// Returns (client_random, session_id, timestamp) on success.
pub fn verify_client_hello(
    data: &[u8],
    secret: &[u8],
) -> Option<([u8; 32], [u8; 32], i64)> {
    if data.len() < 43 {
        return None;
    }
    if data[0] != TLS_RECORD_HANDSHAKE {
        return None;
    }
    if data[5] != 0x01 {
        return None;
    }

    let mut client_random = [0u8; 32];
    client_random.copy_from_slice(&data[CLIENT_RANDOM_OFFSET..CLIENT_RANDOM_OFFSET + CLIENT_RANDOM_LEN]);

    let mut zeroed = data.to_vec();
    zeroed[CLIENT_RANDOM_OFFSET..CLIENT_RANDOM_OFFSET + CLIENT_RANDOM_LEN]
        .copy_from_slice(&[0u8; CLIENT_RANDOM_LEN]);

    let mut mac = HmacSha256::new_from_slice(secret).ok()?;
    mac.update(&zeroed);
    let expected = mac.finalize().into_bytes();

    if !constant_time_eq(&expected[..28], &client_random[..28]) {
        return None;
    }

    let ts_xor: [u8; 4] = [
        client_random[28] ^ expected[28],
        client_random[29] ^ expected[29],
        client_random[30] ^ expected[30],
        client_random[31] ^ expected[31],
    ];
    let timestamp = i64::from_le_bytes([ts_xor[0], ts_xor[1], ts_xor[2], ts_xor[3], 0, 0, 0, 0]);

    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs() as i64;

    if (now - timestamp).abs() > TIMESTAMP_TOLERANCE {
        return None;
    }

    let mut session_id = [0u8; 32];
    if data.len() >= SESSION_ID_OFFSET + SESSION_ID_LEN && data[43] == 0x20 {
        session_id.copy_from_slice(&data[SESSION_ID_OFFSET..SESSION_ID_OFFSET + SESSION_ID_LEN]);
    }

    Some((client_random, session_id, timestamp))
}

/// Build ServerHello response for Fake TLS.
pub fn build_server_hello(
    secret: &[u8],
    client_random: &[u8; 32],
    session_id: &[u8; 32],
) -> Vec<u8> {
    // Template: TLS 1.3 ServerHello
    let mut sh = vec![
        0x16, 0x03, 0x03, 0x00, 0x7a, // record header
        0x02, 0x00, 0x00, 0x76, // handshake type + length
        0x03, 0x03, // version
    ];
    // random placeholder (32 bytes)
    sh.extend_from_slice(&[0u8; 32]);
    sh.push(0x20); // session_id length
    sh.extend_from_slice(session_id); // 32 bytes
    sh.extend_from_slice(&[
        0x13, 0x01, 0x00, // cipher suite
        0x00, 0x2e, // extensions length
        0x00, 0x33, 0x00, 0x24, 0x00, 0x1d, 0x00, 0x20, // key_share
    ]);
    // public key placeholder (32 bytes)
    sh.extend_from_slice(&[0u8; 32]);
    sh.extend_from_slice(&[0x00, 0x2b, 0x00, 0x02, 0x03, 0x04]); // supported_versions

    let sh_random_off = 11;
    let sh_pubkey_off = 89;

    // fill random public key
    let pubkey: [u8; 32] = rand::random();
    sh[sh_pubkey_off..sh_pubkey_off + 32].copy_from_slice(&pubkey);

    let mut response = sh.clone();
    response.extend_from_slice(&CCS_FRAME);

    let encrypted_size: usize = rand::rng().random_range(1900..2100);
    let encrypted_data: Vec<u8> = (0..encrypted_size).map(|_| rand::random()).collect();
    response.push(0x17);
    response.push(0x03);
    response.push(0x03);
    response.extend_from_slice(&(encrypted_size as u16).to_be_bytes());
    response.extend_from_slice(&encrypted_data);

    // HMAC the response with client_random to compute server_random
    let mut mac = HmacSha256::new_from_slice(secret).unwrap();
    mac.update(client_random);
    mac.update(&response);
    let server_random = mac.finalize().into_bytes();

    let mut final_response = response;
    final_response[sh_random_off..sh_random_off + 32].copy_from_slice(&server_random);
    final_response
}

/// Wrap data in TLS application data records.
pub fn wrap_tls_record(data: &[u8]) -> Vec<u8> {
    let mut result = Vec::with_capacity(data.len() + data.len() / TLS_APPDATA_MAX * 5 + 5);
    for chunk in data.chunks(TLS_APPDATA_MAX) {
        result.push(0x17);
        result.push(0x03);
        result.push(0x03);
        result.extend_from_slice(&(chunk.len() as u16).to_be_bytes());
        result.extend_from_slice(chunk);
    }
    result
}

/// Proxy connection to masking domain (when Fake TLS verification fails).
pub async fn proxy_to_masking_domain(
    mut client_reader: tokio::io::ReadHalf<TcpStream>,
    mut client_writer: tokio::io::WriteHalf<TcpStream>,
    initial_data: &[u8],
    domain: &str,
) {
    let up = tokio::time::timeout(
        std::time::Duration::from_secs(10),
        TcpStream::connect((domain, 443)),
    )
    .await;

    let mut up_stream = match up {
        Ok(Ok(s)) => s,
        _ => return,
    };

    STATS.connections_masked.fetch_add(1, std::sync::atomic::Ordering::Relaxed);

    let _ = up_stream.set_nodelay(true);
    if !initial_data.is_empty() {
        let _ = up_stream.write_all(initial_data).await;
        let _ = up_stream.flush().await;
    }

    let (mut up_reader, mut up_writer) = up_stream.split();

    tokio::select! {
        _ = relay(&mut client_reader, &mut up_writer) => {},
        _ = relay(&mut up_reader, &mut client_writer) => {},
    }
}

async fn relay(
    src: &mut (impl AsyncReadExt + Unpin),
    dst: &mut (impl AsyncWriteExt + Unpin),
) {
    let mut buf = [0u8; 16384];
    loop {
        match src.read(&mut buf).await {
            Ok(0) => break,
            Ok(n) => {
                if dst.write_all(&buf[..n]).await.is_err() {
                    break;
                }
                if dst.flush().await.is_err() {
                    break;
                }
            }
            Err(_) => break,
        }
    }
}

/// Wraps a stream to transparently encode/decode TLS Application Data records.
///
/// Reading: strips 5-byte record headers (type 0x17 = AppData), returns raw payload.
/// Writing: wraps data in AppData records before passing to the inner writer.
pub struct FakeTlsStream<R, W> {
    reader: R,
    writer: W,
    buf: VecDeque<u8>,
}

impl<R: AsyncRead + Unpin, W: AsyncWrite + Unpin> FakeTlsStream<R, W> {
    pub fn new(reader: R, writer: W) -> Self {
        Self { reader, writer, buf: VecDeque::new() }
    }

    pub fn into_inner(self) -> (R, W) {
        (self.reader, self.writer)
    }
}

impl<R: AsyncRead + Unpin, W: AsyncWrite + Unpin> AsyncRead for FakeTlsStream<R, W> {
    fn poll_read(
        mut self: Pin<&mut Self>,
        cx: &mut Context<'_>,
        buf: &mut tokio::io::ReadBuf<'_>,
    ) -> Poll<std::io::Result<()>> {
        if !self.buf.is_empty() {
            let n = buf.remaining().min(self.buf.len());
            for b in self.buf.drain(..n) {
                buf.put_slice(&[b]);
            }
            return Poll::Ready(Ok(()));
        }

        let mut hdr = [0u8; 5];
        let mut rb = tokio::io::ReadBuf::new(&mut hdr);
        match Pin::new(&mut self.reader).poll_read(cx, &mut rb) {
            Poll::Ready(Ok(())) if rb.filled().len() < 5 => {
                return Poll::Ready(Err(std::io::Error::new(
                    std::io::ErrorKind::UnexpectedEof,
                    "short TLS record header",
                )));
            }
            Poll::Ready(Err(e)) => return Poll::Ready(Err(e)),
            Poll::Pending => return Poll::Pending,
            _ => {}
        }

        let record_type = hdr[0];
        let len = u16::from_be_bytes([hdr[3], hdr[4]]) as usize;
        let mut body = vec![0u8; len];
        let mut bb = tokio::io::ReadBuf::new(&mut body);
        match Pin::new(&mut self.reader).poll_read(cx, &mut bb) {
            Poll::Ready(Ok(())) if bb.filled().len() < len => {
                return Poll::Ready(Err(std::io::Error::new(
                    std::io::ErrorKind::UnexpectedEof,
                    "short TLS record body",
                )));
            }
            Poll::Ready(Err(e)) => return Poll::Ready(Err(e)),
            Poll::Pending => return Poll::Pending,
            _ => {}
        }

        match record_type {
            0x17 => {
                let n = buf.remaining().min(len);
                buf.put_slice(&body[..n]);
                if n < len {
                    self.buf.extend(&body[n..]);
                }
                Poll::Ready(Ok(()))
            }
            0x14 => {
                self.poll_read(cx, buf)
            }
            _ => Poll::Ready(Err(std::io::Error::new(
                std::io::ErrorKind::InvalidData,
                format!("unexpected TLS record type 0x{:02X}", record_type),
            ))),
        }
    }
}

impl<R: AsyncRead + Unpin, W: AsyncWrite + Unpin> AsyncWrite for FakeTlsStream<R, W> {
    fn poll_write(
        mut self: Pin<&mut Self>,
        cx: &mut Context<'_>,
        buf: &[u8],
    ) -> Poll<Result<usize, std::io::Error>> {
        let wrapped = wrap_tls_record(buf);
        Pin::new(&mut self.writer).poll_write(cx, &wrapped)
    }

    fn poll_flush(
        mut self: Pin<&mut Self>,
        cx: &mut Context<'_>,
    ) -> Poll<Result<(), std::io::Error>> {
        Pin::new(&mut self.writer).poll_flush(cx)
    }

    fn poll_shutdown(
        mut self: Pin<&mut Self>,
        cx: &mut Context<'_>,
    ) -> Poll<Result<(), std::io::Error>> {
        Pin::new(&mut self.writer).poll_shutdown(cx)
    }
}

fn constant_time_eq(a: &[u8], b: &[u8]) -> bool {
    if a.len() != b.len() {
        return false;
    }
    let mut result = 0u8;
    for (x, y) in a.iter().zip(b) {
        result |= x ^ y;
    }
    result == 0
}

File: proxy/fake_tls.py
from __future__ import annotations

import asyncio
import hmac
import hashlib
import os
import random
import struct
import time
import logging

from typing import Optional, Tuple
from .stats import stats


log = logging.getLogger('tg-mtproto-proxy')

TLS_RECORD_HANDSHAKE = 0x16
TLS_RECORD_CCS = 0x14
TLS_RECORD_APPDATA = 0x17

TLS_VERSION_10 = b'\x03\x01'
TLS_VERSION_12 = b'\x03\x03'
TLS_VERSION_13 = b'\x03\x04'

TIMESTAMP_TOLERANCE = 120

TLS_APPDATA_MAX = 16384


_CCS_FRAME = b'\x14\x03\x03\x00\x01\x01'

# --- TLS ClientHello / ServerHello structure offsets ---
# Layout after TLS record header (5 bytes) + hs type+len (3 bytes) + version (2 bytes):
#   [0..31]  random (32 bytes)
#   [32]     session_id_length (1 byte, value 0x20 = 32)
#   [33..64] session_id (32 bytes)
#   [65..67] cipher_suite + compression (3 bytes: 0x13 0x01 0x00)
#   [68..69] extensions_length (2 bytes)
#   [70..73] supported_versions ext header (4 bytes: 0x00 0x33 0x00 0x24)
#   [74..77] key_share ext header (4 bytes: 0x00 0x1d 0x00 0x20)
#   [78..109] key_share public key (32 bytes)
#   [110..115] psk_key_exchange_modes ext (6 bytes: 0x00 0x2b 0x00 0x02 0x03 0x04)

_TLS_RECORD_HDR_LEN = 5      # 0x16 0x03 0x03 <2-byte len>
_TLS_HS_HDR_LEN = 4          # type(1) + len(3)
_TLS_VERSION_LEN = 2         # 0x03 0x03
_TLS_RANDOM_OFFSET = _TLS_RECORD_HDR_LEN + _TLS_HS_HDR_LEN + _TLS_VERSION_LEN  # = 11
_TLS_RANDOM_LEN = 32
_TLS_SESSION_ID_LEN_OFFSET = _TLS_RANDOM_OFFSET + _TLS_RANDOM_LEN             # = 43
_TLS_SESSION_ID_OFFSET = _TLS_SESSION_ID_LEN_OFFSET + 1                       # = 44
_TLS_SESSION_ID_LEN = 32
_TLS_CIPHERSUITE_LEN = 3     # 0x13 0x01 0x00
_TLS_EXT_LEN_OFFSET = _TLS_SESSION_ID_OFFSET + _TLS_SESSION_ID_LEN + _TLS_CIPHERSUITE_LEN  # = 79
_TLS_EXT_HEADER_LEN = 2
_TLS_SUPPORTED_VERSIONS_OFF = _TLS_EXT_LEN_OFFSET + _TLS_EXT_HEADER_LEN       # = 81
_TLS_SUPPORTED_VERSIONS_HDR_LEN = 4  # 0x00 0x33 0x00 0x24
_TLS_KEY_SHARE_EXT_OFF = _TLS_SUPPORTED_VERSIONS_OFF + _TLS_SUPPORTED_VERSIONS_HDR_LEN  # = 85
_TLS_KEY_SHARE_HDR_LEN = 4   # 0x00 0x1d 0x00 0x20
_TLS_PUBKEY_OFF = _TLS_KEY_SHARE_EXT_OFF + _TLS_KEY_SHARE_HDR_LEN             # = 89
_TLS_PUBKEY_LEN = 32

_TLS_SH_RANDOM_OFF = _TLS_RANDOM_OFFSET     # = 11  (same layout as ClientHello)
_TLS_SH_SESSID_OFF = _TLS_SESSION_ID_OFFSET # = 44
_TLS_SH_PUBKEY_OFF = _TLS_PUBKEY_OFF        # = 89

_SERVER_HELLO_TEMPLATE = bytearray(
    b'\x16\x03\x03\x00\x7a'            # TLS record header
    b'\x02\x00\x00\x76'                # ServerHello type + length
    b'\x03\x03'                         # TLS 1.2
    + b'\x00' * _TLS_RANDOM_LEN        # server random
    + bytes([_TLS_SESSION_ID_LEN])      # session_id_length = 32
    + b'\x00' * _TLS_SESSION_ID_LEN    # session_id
    + b'\x13\x01\x00'                  # cipher suite TLS_AES_128_GCM + null compression
    + b'\x00\x2e'                      # extensions length
    + b'\x00\x33\x00\x24\x00\x1d\x00\x20'  # supported_versions + key_share headers
    + b'\x00' * _TLS_PUBKEY_LEN        # ephemeral public key
    + b'\x00\x2b\x00\x02\x03\x04'      # psk_key_exchange_modes
)


def verify_client_hello(data: bytes, secret: bytes) -> Optional[Tuple[bytes, bytes, int]]:
    n = len(data)
    # Minimum: 5 (record hdr) + 4 (hs type+len) + 2 (version) + 32 (random) = 43
    min_random_end = _TLS_RANDOM_OFFSET + _TLS_RANDOM_LEN
    if n < min_random_end:
        return None
    if data[0] != TLS_RECORD_HANDSHAKE:
        return None
    if data[5] != 0x01:
        return None

    client_random = bytes(data[_TLS_RANDOM_OFFSET:_TLS_RANDOM_OFFSET + _TLS_RANDOM_LEN])

    zeroed = bytearray(data)
    zeroed[_TLS_RANDOM_OFFSET:_TLS_RANDOM_OFFSET + _TLS_RANDOM_LEN] = b'\x00' * _TLS_RANDOM_LEN

    expected = hmac.new(secret, bytes(zeroed), hashlib.sha256).digest()

    if not hmac.compare_digest(expected[:28], client_random[:28]):
        return None

    ts_xor = bytes(client_random[28 + i] ^ expected[28 + i] for i in range(4))
    timestamp = struct.unpack('<I', ts_xor)[0]

    now = int(time.time())
    if abs(now - timestamp) > TIMESTAMP_TOLERANCE:
        return None

    session_id = b'\x00' * _TLS_SESSION_ID_LEN
    if n >= _TLS_SESSION_ID_OFFSET + _TLS_SESSION_ID_LEN and data[_TLS_SESSION_ID_LEN_OFFSET] == _TLS_SESSION_ID_LEN:
        session_id = bytes(data[_TLS_SESSION_ID_OFFSET:_TLS_SESSION_ID_OFFSET + _TLS_SESSION_ID_LEN])

    return client_random, session_id, timestamp


def build_server_hello(secret: bytes, client_random: bytes, session_id: bytes) -> bytes:
    sh = bytearray(_SERVER_HELLO_TEMPLATE)
    sh[_TLS_SH_SESSID_OFF:_TLS_SH_SESSID_OFF + _TLS_SESSION_ID_LEN] = session_id
    sh[_TLS_SH_PUBKEY_OFF:_TLS_SH_PUBKEY_OFF + _TLS_PUBKEY_LEN] = os.urandom(_TLS_PUBKEY_LEN)

    ccs = _CCS_FRAME
    encrypted_size = random.randint(1900, 2100)
    encrypted_data = os.urandom(encrypted_size)
    app_record = b'\x17\x03\x03' + struct.pack('>H', encrypted_size) + encrypted_data

    response = bytes(sh) + ccs + app_record

    hmac_input = client_random + response
    server_random = hmac.new(secret, hmac_input, hashlib.sha256).digest()

    final = bytearray(response)
    final[_TLS_SH_RANDOM_OFF:_TLS_SH_RANDOM_OFF + _TLS_RANDOM_LEN] = server_random

    return bytes(final)


def wrap_tls_record(data: bytes) -> bytes:
    parts = []
    offset = 0
    while offset < len(data):
        chunk = data[offset:offset + TLS_APPDATA_MAX]
        parts.append(
            b'\x17\x03\x03'
            + struct.pack('>H', len(chunk))
            + chunk
        )
        offset += len(chunk)
    return b''.join(parts)


class FakeTlsStream:
    __slots__ = ('_reader', '_writer', '_read_buf', '_read_left')

    def __init__(self, reader: asyncio.StreamReader, writer: asyncio.StreamWriter):
        self._reader = reader
        self._writer = writer
        self._read_buf = bytearray()
        self._read_left = 0

    async def readexactly(self, n: int) -> bytes:
        while len(self._read_buf) < n:
            payload = await self._read_tls_payload()
            if not payload:
                raise asyncio.IncompleteReadError(bytes(self._read_buf), n)
            self._read_buf.extend(payload)
        result = bytes(self._read_buf[:n])
        del self._read_buf[:n]
        return result

    async def read(self, n: int) -> bytes:
        if self._read_buf:
            chunk = bytes(self._read_buf[:n])
            del self._read_buf[:n]
            return chunk
        payload = await self._read_tls_payload()
        if not payload:
            return b''
        if len(payload) > n:
            self._read_buf.extend(payload[n:])
            return payload[:n]
        return payload

    async def _read_tls_payload(self) -> bytes:
        if self._read_left > 0:
            data = await self._reader.read(self._read_left)
            if not data:
                return b''
            self._read_left -= len(data)
            return data

        while True:
            hdr = await self._reader.readexactly(5)
            rtype = hdr[0]
            rec_len = struct.unpack('>H', hdr[3:5])[0]

            if rtype == TLS_RECORD_CCS:
                if rec_len > 0:
                    await self._reader.readexactly(rec_len)
                continue

            if rtype != TLS_RECORD_APPDATA:
                return b''

            data = await self._reader.read(min(rec_len, 65536))
            if not data:
                return b''
            remaining = rec_len - len(data)
            if remaining > 0:
                self._read_left = remaining
            return data

    def write(self, data: bytes) -> None:
        self._writer.write(wrap_tls_record(data))

    async def drain(self) -> None:
        await self._writer.drain()

    def close(self) -> None:
        self._writer.close()

    async def wait_closed(self) -> None:
        await self._writer.wait_closed()

    def get_extra_info(self, name, default=None):
        return self._writer.get_extra_info(name, default)

    @property
    def transport(self):
        return self._writer.transport

    def is_closing(self):
        return self._writer.is_closing()


async def proxy_to_masking_domain(reader, writer, initial_data: bytes,
                                    domain: str, label: str) -> None:
    try:
        up_reader, up_writer = await asyncio.wait_for(
            asyncio.open_connection(domain, 443), timeout=10)
    except Exception as exc:
        log.warning("[%s] masking: cannot connect to %s:443: %s",
                  label, domain, repr(exc))
        return

    log.debug("[%s] masking -> %s:443", label, domain)
    stats.connections_masked += 1

    try:
        if initial_data:
            up_writer.write(initial_data)
            await up_writer.drain()

        async def _relay(src, dst):
            try:
                while True:
                    chunk = await src.read(16384)
                    if not chunk:
                        break
                    dst.write(chunk)
                    await dst.drain()
            except (ConnectionResetError, BrokenPipeError, OSError,
                    asyncio.CancelledError):
                pass
            finally:
                try:
                    dst.close()
                    await dst.wait_closed()
                except Exception:
                    pass

        await asyncio.gather(
            _relay(reader, up_writer),
            _relay(up_reader, writer),
        )
    except Exception:
        pass
    finally:
        try:
            up_writer.close()
        except Exception:
            pass
